modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/impl/TestMvPartitionStorage.java

```diff
@@ -204,7 +215,18 @@ public void commitWrite(RowId rowId, HybridTimestamp timestamp) {
                 return versionChain;
             }
 
-            return VersionChain.forCommitted(timestamp, versionChain);
+            VersionChain committedVersionChain = VersionChain.forCommitted(timestamp, versionChain);
+
+            if (committedVersionChain.next != null) {
+                // Avoid creating tombstones for tombstones.
+                if (committedVersionChain.row == null && committedVersionChain.next.row == null) {
+                    return committedVersionChain.next;
+                }
+
+                gcQueue.add(new IgniteBiTuple<>(committedVersionChain, rowId));
```

Comment:
`ConcurrentSkipListMap#compute()` may invoke the remapping function more than once due to concurrency. Here, the remapping function has a side effect (addition to the queue). It seems that it can result in multiple insertion to the queue. `VersionChain` does not define `equals()/hashCode()`, so this will really result in a chain being stored twice in the GC queue.

Reply:
Also, access to `gcQueue` is not synchronized explicitly. `ConcurrentMap` guarantees happens-before between actions with the map and actions around these actions (before or after them in the Program Order), but does it make any guarantees about what happens inside remapping functions? Documentation on `ConcurrentSkipListMap#compute()` specifically says that it does NOT guarantee atomicity for remapping functions.

Reply:
My bad, I'll fix it

Reply:
You know what, gcQueue is defined with a custom comparator, right? I guess equality is not an issue and hashCode is never called, so writing to the queue several times should be fine.

Reply:
That is true, comparator should remove this '2 entries' problem. But there is still '2 writes to the queue' problem.

Imagine that:

1. `map.compute()` makes first attempt, adds an entry to the queue
2. GC thread calls `pollForVacuum()` and removes this entry
3. `map.compute()` makes a second attempt and adds a second entry (equal to the first one) for the second time
4. GC thread polls again, so it sees the same (actually, a different, but equal) entry for the second time

Is this possible? Is this valid?

Reply:
GC cannot remove this entry, we only remove "old" entries, and even they must have a "parent" in the chain. Situation that you're describing is only possible in tests, and I don't see a necessity to cover it

Reply:
Continuing the example. We now have 2 identical entries in the GC queue. When the watermark is high enough, GC grabs first of them, nullifies the pointer to the version it shadows, but this changes nothing from the point of view of the user, because this entry is not included in the chain. So GC vacuumed something, but in reality it did not. This might cause confusing failures in tests.

Why don't we just add synchronization to make sure it behaves as predictable as possible? This is a test storage, after all. If we must have a trade-off between performance and simplicity/predictability, we should choose the latter.

Reply:
> We now have 2 identical entries in the GC queue.

There can be no duplicates in the queue. Anyway, I think I found a different example where the map itself can break regardless of the queue.

> Why don't we just add synchronization to make sure it behaves as predictable as possible?

I think I will